src/Streamly/Internal/Data/Array/Stream/Foreign.hs

```diff
@@ -480,6 +553,105 @@ parseD (PRD.Parser pstep initial extract) stream@(D.Stream step state) = do
 
     where
 
+    -- "backBuf" contains last few items in the stream that we may have to
+    -- backtrack to.
+    --
+    -- XXX currently we are using a dumb list based approach for backtracking
+    -- buffer. This can be replaced by a sliding/ring buffer using Data.Array.
+    -- That will allow us more efficient random back and forth movement.
+    {-# INLINE go #-}
+    go !_ st backBuf !pst = do
+        r <- step defState st
+        case r of
+            D.Yield (Array (ForeignPtr start contents) (Ptr end)) s ->
+                gobuf SPEC s backBuf (ForeignPtr end contents) (Ptr start) pst
+            D.Skip s -> go SPEC s backBuf pst
+            D.Stop -> do
+                b <- extract pst
+                return (b, D.nil)
+
+    -- Use strictness on "cur" to keep it unboxed
+    gobuf !_ s backBuf fp@(ForeignPtr end _) !cur !pst
+        | cur == Ptr end = do
+            liftIO $ touchForeignPtr fp
+            go SPEC s backBuf pst
+    gobuf !_ s backBuf fp@(ForeignPtr end contents) !cur !pst = do
+        x <- liftIO $ peek cur
+        pRes <- pstep pst x
+        let elemSize = sizeOf (undefined :: a)
+            next = cur `plusPtr` elemSize
+        case pRes of
+            PR.Partial 0 pst1 ->
+                 gobuf SPEC s (List []) fp next pst1
+            PR.Partial n pst1 -> do
+                assert (n <= Prelude.length (x:getList backBuf)) (return ())
+                let src0 = Prelude.take n (x:getList backBuf)
+                    arr0 = A.fromListN n (Prelude.reverse src0)
+                    !(Ptr curAddr) = next
+                    arr1 = Array (ForeignPtr curAddr contents) (Ptr end)
+                    src = arr0 <> arr1
```

Comment:
You could probably save the arrays in the `backBuf` instead.

Reply:
Never mind, this looks good. Need to think about extensive backtracking where you need to append arrays multiple times triggering a lot of `realloc`